### PR TITLE
Use TemporaryDirectory instead of mkdtemp in a few places.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

(main remaining place is backend_pgf which has a rather intricate way to set up the tempdir cleanup that'll need to be unwound at some point)

Edit: Also use TemporaryDirectory instead of mkstemp in backend_ps.py; the reason to use that instead of (Named)TemporaryFile is to make it easy to have a closable temporary file on Windows (where a temporary file cannot be opened both by python and by an external process (gs) at the same time).

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_bases.py [hunk 1]
```diff
@@ -50,30 +49,22 @@ def check(master_transform, paths, all_transforms,
...
-        test_dir = tempfile.mkdtemp()
-        plt.rcParams['savefig.directory'] = test_dir
-        fig = plt.figure()
-        canvas = FigureCanvasBase(fig)
-        filename = canvas.get_default_filename()
-        assert filename == 'image.png'
-    finally:
-        shutil.rmtree(test_dir)
+def test_get_default_filename(tmpdir):
+    plt.rcParams['savefig.directory'] = tmpdir
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py]: `str(tmpdir)`

File: lib/matplotlib/tests/test_backend_bases.py [hunk 2]
```diff
@@ -50,30 +49,22 @@ def check(master_transform, paths, all_transforms,
...
 
-def test_get_default_filename_already_exists():
+def test_get_default_filename_already_exists(tmpdir):
     # From #3068: Suggest non-existing default filename
-    try:
-        test_dir = tempfile.mkdtemp()
-        plt.rcParams['savefig.directory'] = test_dir
-        fig = plt.figure()
-        canvas = FigureCanvasBase(fig)
+    plt.rcParams['savefig.directory'] = tmpdir
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py]: `str(tmpdir)`

File: lib/matplotlib/tests/test_backend_bases.py [hunk 3]
```diff
@@ -50,30 +49,22 @@ def check(master_transform, paths, all_transforms,
...
+    fig = plt.figure()
+    canvas = FigureCanvasBase(fig)
 
-        # create 'image.png' in figure's save dir
-        open(os.path.join(test_dir, 'image.png'), 'w').close()
+    # create 'image.png' in figure's save dir
+    open(os.path.join(tmpdir, 'image.png'), 'w').close()
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py]: `tmpdir.join('image.png').ensure()` (I think)

Comment (anntzer) [reply lib/matplotlib/tests/test_backend_bases.py]: not convinced it's worth using the pytest-specific api here.
fixed other points.

File: lib/matplotlib/tests/test_style.py [hunk 4]
```diff
@@ -23,21 +25,16 @@ def temp_style(style_name, settings=None):
...
-    # Add temp directory to style path and reload so we can access this style.
-    USER_LIBRARY_PATHS.append(tempdir)
-    style.reload_library()
-
     try:
-        yield
+        with TemporaryDirectory() as tmpdir:
+            # Write style settings to file in the tmpdir.
+            Path(tmpdir, temp_file).write_text(
+                "\n".join(map("{0[0]}: {0[1]}".format, settings.items())))
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_style.py]: I'd rather go with `"\n".join("{}: {}".format(k, v) for k, v in settings.items())`, which is more readable than mapping and tuple-item access in format.

Comment (anntzer) [reply lib/matplotlib/tests/test_style.py]: sure

**Pull request merged**